src/sage/schemes/projective/coherent_sheaf.py

```diff
@@ -0,0 +1,323 @@
+        representation is computed.
+
+        INPUT:
+
+        - ``r`` -- (default: 0) a non-negative integer
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: s = X.structure_sheaf()
+            sage: s._cohomology_group(1)
+            Vector space quotient V/W of dimension 3 over Rational Field where
+            V: Vector space of degree 3 and dimension 3 over Rational Field
+            Basis matrix:
+            [1 0 0]
+            [0 1 0]
+            [0 0 1]
+            W: Vector space of degree 3 and dimension 0 over Rational Field
+            Basis matrix:
+            []
+            sage: s._cohomology_group(1).dimension()  == s.cohomology(1)
+            True
+        """
+        return self._cohomology.H(r)
+
+    def cohomology(self, r=0):
+        """
+        Return the dimension of the `r`-th cohomology as a vector space.
+
+        INPUT:
+
+        - ``r`` -- (default: 0) a non-negative integer
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sheaf = X.structure_sheaf()
+            sage: sheaf.cohomology(0)
+            1
+            sage: sheaf.cohomology(1)
+            3
+            sage: sheaf.cohomology(2)
+            0
+        """
+        return self._cohomology.h(r)
+
+    def twist(self, t=0):
+        r"""
+        Return the twisted sheaf `F(n)` of this sheaf `F`.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sh = X.structure_sheaf()
+            sage: sh.twist(1).euler_characteristic()
+            2
+            sage: sh.twist(2).euler_characteristic()
+            6
+        """
+        return type(self)(self._base_scheme, self._module, self._twist + t)
+
+    def euler_characteristic(self):
+        """
+        Return the Euler characteristic of this coherent sheaf.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sh = X.structure_sheaf()
+            sage: sh.euler_characteristic()
+            -2
+        """
+        d = self._base_scheme.dimension()
+        chi = 0
+        for r in range(d + 1):  # for Grothendieck's vanishing theorem
+            d = self.cohomology(r)
+            if r % 2:
+                chi = chi - d
+            else:
+                chi = chi + d
+        return chi
+
+
+class CoherentSheaf_on_projective_space(CoherentSheaf):
+    r"""
+    Coherent sheaf on a projective space.
+
+    EXAMPLES::
+
+        sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+        sage: P2.structure_sheaf()
+        Coherent sheaf on Projective Space of dimension 2 over Rational Field
+    """
+    @cached_property
```

Comment:
What `@cached_property` instead of `@lazy_attribute`? Also, wouldn't it make more sense for `MaruyamaComplex` to construct itself lazily?

Reply:
`@cached_property` is standard python while `@lazy_attribute` is a sage thing. As far as I know, there is no difference in functionality. Is there?

I prefer `@cached_property` because I am aware that using certain sage-specific decorators is presently an obstacle in removing `src/sage/sage_docbuild/ext/sage_autodoc.py`, which is preventing us to use vanilla Sphinx and is a major friction in upgrading Sphinx. I hope someday to replace all `@lazy_attribute` with `@cached_property` :-)

`MaruyamaComplex` is an internal class. I don't see a need to make it fancy.

Reply:
It is better design to have `MaruyamaComplex` be a `UniqueRepresentation`. This will handle the appropriate caching, it could then be pickled (so used in multiprocessing, used across sessions, etc.), and it's not internal since you give the user access to it (which I think is the right thing to do).

re: `@cached_property`, I am not sure about the differences between the two. From some basics checks, they agree, but I didn't check carefully or in Cython. I would be *extremely* careful about such mass replacements at this point.

Reply:
> It is better design to have `MaruyamaComplex` be a `UniqueRepresentation`. This will handle the appropriate caching, it could then be pickled (so used in multiprocessing, used across sessions, etc.), 

Appropriate caching is handled by  `@cached_property`, and there is no problem in pickling.

```
sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
sage: X = P2.subscheme([x^4 + y^4 + z^4])
sage: sh = X.structure_sheaf()
sage: sh
Coherent sheaf on Closed subscheme of Projective Space of dimension 2 over Rational Field defined by:
  x^4 + y^4 + z^4
sage: loads(dumps(sh))
Coherent sheaf on Closed subscheme of Projective Space of dimension 2 over Rational Field defined by:
  x^4 + y^4 + z^4
sage: loads(dumps(sh._cohomology))
Maruyama Method using S(0) <-- S(-4) <-- 0
```

I don't see a benefit of using `UniqueRepresentation`...

> and it's not internal since you give the user access to it (which I think is the right thing to do).

There is user access via `sh._cohomology`. It is still internal. What do you think is not internal?

Reply:
I removed `_cohomology_group()` method of sheaf, by the way.